Automattic/wp-calypso
# Composite Checkout: Simplify API by adding CheckoutStepGroup, CheckoutFormSubmit, and PaymentMethodStep

#### Changes proposed in this Pull Request

The `@automattic/composite-checkout` package has been around for a while now and its API has changed considerably from its initial design. The result is that its documentation is difficult to read and trying to understand how to use it is quite complex. It has a massive number of exports, but most of them are only useful in very specific circumstances.

This PR attempts to simplify the API with as few breaking changes as possible. Specifically:

- Rather than needing `Checkout` to wrap the whole contents and `CheckoutStepArea` and/or `CheckoutSteps` to wrap steps, there is now one wrapper, `CheckoutStepGroup`, which can handle both steps and non-steps.
- Rather than passing submit button props to `CheckoutStepArea`, those are passed directly to a new component for the submit button, `CheckoutFormSubmit`.
- To create a payment method step, rather than needing to call `getDefaultPaymentMethodStep()` and then pass its return values to `CheckoutStep` manually, there is a simpler `PaymentMethodStep` component.

The breaking changes are as follows:

- `CheckoutStepArea` no longer automatically adds a `CheckoutSubmitButton`; that must be done explicitly by inserting `CheckoutFormSubmit`. This PR adds the new component where it is needed and removes all public uses of `CheckoutStepArea`.
- `Checkout` no longer has a set of default steps if no children are provided. This feature was never used and I don't believe it was ever mentioned in the documentation.

One other significant change that hopefully won't be noticable is that `CheckoutStepArea` previously had [a large amount of logic](https://github.com/Automattic/wp-calypso/blob/d203a8638a2debd5078dc0eb81ab11907cbad2fd/packages/composite-checkout/src/components/checkout-steps.tsx#L502-L539) dedicated to leaving sufficient room for the footer which was added in https://github.com/Automattic/wp-calypso/pull/54169. The logic wasn't easy to retain because the submit button is no longer part of that component, so I've replaced it with a new private hook, `useCustomPropertyForHeight()`, that uses CSS custom properties to change margins. This is more fragile since it relies on a custom property to pass the margin along, but it seemed like the easiest way to pass styles between disconnected components.

#### Testing instructions

First, it would be a good idea to go through [the updated README](https://github.com/Automattic/wp-calypso/blob/f5ebb73cdde798df7380335650a97517da9ef8fa/packages/composite-checkout/README.md) and see if its simplified description is comprehensive.

Second, prior to this PR, the composite-checkout demo was broken for quite a while. Here we fix it so you can run that and try it out: `yarn workspace @automattic/composite-checkout run storybook`

Third, this modifies calypso checkout, so it's a good idea to try that out, although most of the important behavior is covered by automated tests. Add a product to the cart and visit checkout, then click around between steps to make sure they change correctly, and that the view looks reasonable at different widths.

In particular, verify that the checkout total is fully visible when scrolling to the bottom of the page at small widths (make sure the submit button does not overlap it).

No other parts of calypso should be affected by this PR.
